File: week_8_and_9/4_Ruby/fizz_buzz/my_solution.rb
# U3.W8-9: 


# I worked on this challenge [by myself].

# 2. Pseudocode

=begin
iterate through the array
make empty array
	if i%15 == 0
		shovel FizzBuzz
	elsif i%5 == 0
		shovel Buzz
	elsif i%3 == 0	
		shovel Fizz
	else
		shovel i 
=end


# 3. Initial Solution

# def super_fizzbuzz(array)
# supered = []
# 	array.each do |x|
# 		if x % 15 == 0
# 			supered.push("FizzBuzz")
# 		elsif x % 5 == 0
# 			supered.push("Buzz")
# 		elsif x % 3 == 0
# 			supered.push("Fizz")
# 		else
# 			supered.push(x)
# 		end		
# 	end	
# 	return supered
# end



# 4. Refactored Solution

def super_fizzbuzz(array)
supered = []
	array.each do |x|
		if x % 15 == 0
			supered << "FizzBuzz"
		elsif x % 5 == 0
			supered << "Buzz"
		elsif x % 3 == 0
			supered << "Fizz"
		else
			supered << x
		end		
	end	
	return supered
end




# 1. DRIVER TESTS/ASSERT STATEMENTS GO BELOW THIS LINE


def assert
	raise "Assertion failed!" unless yield
end
assert { super_fizzbuzz([1,2,3])  == [1, 2, "Fizz"]}
assert { super_fizzbuzz([1,2,5])  == [1, 2, "Buzz"]}
assert { super_fizzbuzz([1,2,15]) == [1, 2, "FizzBuzz"]}
assert { super_fizzbuzz([30, 9, 20, 1]) == ["FizzBuzz", "Fizz", "Buzz", 1]}



# 5. Reflection 
=begin
What parts of your strategy worked? What problems did you face?
I basically thought the problem out and pseudocoded and then coded the solution.  At first I was trying to modify the original
array and that didn't work quite as I hoped, so I moved to creating a new super-ed array and that worked a lot more smoothly.
I didn't really face major problems outside that though.


What questions did you have while coding? What resources did you find to help you answer them?
I didn't really have any major questions while coding, but I referenced ruby docs a few times fora refresher on arrays.


What concepts are you having trouble with, or did you just figure something out? If so, what?
The logic of the problem (order of the if statements) wasn't bad, but the implementation tripped me up a bit.  I didn't have
too much trouble though.


Did you learn any new skills or tricks?
Not really, considering this week is review.


How confident are you with each of the Learning Competencies?
I think I'm alright with the learning competencies this time!


Which parts of the challenge did you enjoy?
I enjoyed thinking through the problem and getting the code to work


Which parts of the challenge did you find tedious?
None
=end

File: week_8_and_9/3_JavaScript/grocery_list/my_solution.js
// U3.W8-9: 


// I worked on this challenge [by myself].

// 2. Pseudocode
/*
create class grocery list (in this case a function)
create empty array list
    
create add item method
    input is item
    push item into list
    
create delete item method
    input is item
    delete from list
        
create view list method
    no input
    use console log to print
    checked list
    unchecked list

create class grocery item (in this case a function)
    takes 2 inputs, item name and quantity

*/


// 3. Initial Solution

// function List() {}

// var list = [];

// List.prototype.addItem = function(item){
//     list.push(item);
// };

// List.prototype.deleteItem = function(item){
//     var dindex = list.indexOf(item);
//     if (dindex > -1) {
//         list.splice(dindex, 1);
//     }
// };

// List.prototype.viewList = function(){
//     console.log("GROCERY LIST");
//     for (var i = 0; i < list.length; i++){
//         console.log(list[i]);
//     }
// };

// function Item(item, quantity) { 
//     this.item = item;
//     this.quantity= quantity;
//     }

    
// 4. Refactored Solution

function List() {}

var list = [];

List.prototype.addItem = function(item){
    list.push(item);
};

List.prototype.deleteItem = function(item){
    var index = list.indexOf(item);
    if (index > -1) {
        list.splice(index, 1);
    }
};

List.prototype.viewList = function(){
    console.log("GROCERY LIST");
    for (var i = 0; i < list.length; i++){
        console.log(list[i]);
    }
};

function Item(item, quantity) { 
    this.item = item;
    this.quantity= quantity;
    }




// 1. DRIVER TESTS/ASSERT STATEMENTS GO BELOW THIS LINE
var apple = new Item("apple", 3);
var rice = new Item("rice", 1);
var beef = new Item("beef", 2);
var glist = new List();

glist.addItem(apple);
glist.addItem(rice);
glist.addItem(beef);
glist.deleteItem(beef);
glist.viewList();





// 5. Reflection 

// What parts of your strategy worked? What problems did you face?
// 	I pseudocoded, then when I was implementing the code I realized I wasn't totally sure how to do most of it in javascript.
// 	I also originally included more features then I really needed to which made it more complicated.

// What questions did you have while coding? What resources did you find to help you answer them?
// 	I had questions about object oriented javascript, and I referenced this source
// 	https://developer.mozilla.org/en-US/docs/Web/JavaScript/Introduction_to_Object-Oriented_JavaScript

// 	I also followed a previous suggestion from a classmate to use http://repl.it/ to test more easily.

// What concepts are you having trouble with, or did you just figure something out? If so, what?
// 	I'm having trouble with javascript because I feel like I have to stop and ook up a lot of my questions frequently.
// 	I'm still not confident about object oriented javascript.

// Did you learn any new skills or tricks?
// 	I read up a bit on object oriented javascript.  And while that isn't exactly new, I did learn a bit.

// How confident are you with each of the Learning Competencies?
// 	Not too confident.  I was about to ask for help from cohort mates because I was getting frustrated, and then I figured
// 	out that I was using the wrong syntax for something

// Which parts of the challenge did you enjoy?
// 	I enjoyed pseudocoding and breaking the challenge down into smaller steps

// Which parts of the challenge did you find tedious?
// 	Looking up javascript questions/syntax constantly.